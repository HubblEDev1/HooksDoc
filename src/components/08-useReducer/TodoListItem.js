import React from 'react'
import './styles.css';

export const TodoListItem = ({item, id, handleDelete, handleToggle}) => {
    return (
        <li 
            key={item.id}
            className = "list-group-item"
        >
            <p 
                className={`${item.done && 'complete'}`} 
                onClick = {() => handleToggle(item.id)}
            >
                {id + 1}. {item.desc}
            </p>
            <button 
                className="btn btn-danger"
                onClick={() => handleDelete(item.id)}
            >
                Borrar
            </button>
        </li>
    )
}
