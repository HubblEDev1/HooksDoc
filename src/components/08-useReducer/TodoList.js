import React from 'react'
import './styles.css';
import { TodoListItem } from './TodoListItem';


export const TodoList = ({state, handleDelete, handleToggle}) => {
    return (
        <ul className="list-group list-group-flush">
            {
                state.map(( item, i ) => (
                    <TodoListItem key={item.id} item={item} id={i} handleDelete={handleDelete} handleToggle={handleToggle}/>
                    )
                )
            }
        </ul>
    )
}
