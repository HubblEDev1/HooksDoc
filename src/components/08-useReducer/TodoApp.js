import React, { useEffect, useReducer } from 'react'
import { todoReducer } from './todoReducer';

import './styles.css';
import { useForm } from '../../hooks/useForm';
import { TodoList } from './TodoList';
import { TodoAdd } from './TodoAdd';

const init = () => {
    
    return JSON.parse(localStorage.getItem('state')) || [];
    
    /*return [{
        id: new Date().getTime(),
        desc: 'BMW',
        done: false
    }];*/
};

export const TodoApp = () => {
    const [state, dispatch] = useReducer(todoReducer, [], init);//init is the initialState an by default the initiState is [+


    useEffect(() => {
        localStorage.setItem('state', JSON.stringify(state))
    }, [state])

    const handleDelete = (todoId) => {
        console.log(todoId);
        const action = {
            type: 'delete',
            payload: todoId
        }

        dispatch(action)
        
    }

    const handleToggle = (todoId) => {
        dispatch({
            type: 'toggle',
            payload: todoId
        })

    }

    const handleAddTodo = ( newTodo ) => {
        dispatch({
            type: 'add',
            payload: newTodo
        });
    }

    return (
        <div>
            <h1>Todo List ({state.length})</h1>
            <hr/>

            <div className="row">
                <div className="col-7"> 
                    todos
                    <TodoList state={state} handleDelete={handleDelete} handleToggle={handleToggle}/>
                </div>
                <div className="col-5">
                    <TodoAdd handleAddTodo={handleAddTodo}/>
                </div>
            </div>
        </div>
    )
}
