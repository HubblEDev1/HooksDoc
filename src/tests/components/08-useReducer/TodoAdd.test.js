import { shallow } from "enzyme";
import { TodoAdd } from "../../../components/08-useReducer/TodoAdd";

describe('Testing todoList', () => {
    const handleAddTodo = jest.fn();
    
    const wrapper = shallow(<TodoAdd
        handleAddTodo = {handleAddTodo}    
    />);
    
    test('Take snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('Not call handleAddTodo', () => {
        const formSubmit = wrapper.find('form').prop('onSubmit');
        //console.log(wrapper.find('form').prop('onSubmit'));
        formSubmit({preventDefault(){}});

        expect(handleAddTodo).toHaveBeenCalledTimes(0);
    });

    test('call handleAddTodo', () => {
        const input = wrapper.find('input');
        
        const value = 'Learn ...';

        //1.- Simular el inputChange,
        input.simulate('change', {target: {
            value,
            name: 'description'
        }});
        
        //2.- Simular el submit 
        const formSubmit = wrapper.find('form').prop('onSubmit');
        formSubmit({preventDefault(){}});        
        //console.log(wrapper.find('form').prop('onSubmit'));
        //formSubmit({preventDefault(){}});
        expect(handleAddTodo).toHaveBeenCalledTimes(1);
        expect(handleAddTodo).toHaveBeenCalledWith(expect.any(Object));
        expect(handleAddTodo).toHaveBeenCalledWith({
            id: expect.any(Number),
            desc: value,
            done: false
        });   
        expect(wrapper.find('input').prop('value')).toBe(''); 
    });
})