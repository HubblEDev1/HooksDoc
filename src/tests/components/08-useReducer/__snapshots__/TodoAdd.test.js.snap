// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing todoList Take snapshot 1`] = `
<Fragment>
  <h4>
    Agregar TODO
  </h4>
  <hr />
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="form-control"
      name="description"
      onChange={[Function]}
      placeholder="Aprender..."
      type="text"
      value=""
    />
    <button
      className="btn btn-outline-primary mt-1 btn-block"
    >
      Add
    </button>
  </form>
</Fragment>
`;
